Add an AutogradMeta factory.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28620 Merge Tensor and Variable.
* #28610 Null AutogradMeta optimization
* #28609 Delete dead TensorImpl::detach_autograd_meta
* #28602 Remove unused gradient_edge argument from make_variable_view
* **#28593 Add an AutogradMeta factory.**
* #28601 Unfriend Variable factory functions.
* #28592 Move all autograd_meta_ manipulating operations out-of-line.
* #28543 Test TensorTypeSet instead of autograd_meta_ for variable-ness.

When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library. So I need another virtual factory interface
to be able to do this.

I also adjust the AutogradMeta constructor so that the TensorImpl argument is
optional. This argument is only needed if `requires_grad == True`, as we use it
to test if the variable is valid (only floating point tensors can have requires grad true).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171161](https://our.internmc.facebook.com/intern/diff/D18171161)

ezyang committed:
Add an AutogradMeta factory.

When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add an AutogradMeta factory."


When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add an AutogradMeta factory."


When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add an AutogradMeta factory."


When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library. So I need another virtual factory interface
to be able to do this.

I also adjust the AutogradMeta constructor so that the TensorImpl argument is
optional. This argument is only needed if `requires_grad == True`, as we use it
to test if the variable is valid (only floating point tensors can have requires grad true).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

yf225 commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -28,7 +28,10 @@ AutogradMeta::AutogradMeta(at::TensorImpl* self_impl, bool requires_grad, Edge g
-  set_requires_grad(requires_grad, self_impl);
+  if (requires_grad) {
+    TORCH_INTERNAL_ASSERT(self_impl);
+    set_requires_grad(requires_grad, self_impl);
```
Do we assume that `set_requires_grad` is always no-op when `requires_grad` is false? (Not sure if this is the semantics we are moving to)

ezyang replied:
In general, `set_requires_grad` is not a no-op when `requires_grad`, but in this particular case, as the factory function for AutogradMeta, it is indeed a no-op, because the default setting for `requires_grad` false.

yf225 replied:
When a Variable has AutogradMeta and the user calls `set_requires_grad(false)`, do we set the AutogradMeta to nullptr again? (Is this logic handled elsewhere?)

ezyang replied:
We discussed this offline, but as this patch stands right now, the answer is no (because you have to do a consistency check on the rest of AutogradMeta to see if it matches the default constructed form). I agreed to add a comment about this.

ezyang committed:
Update on "Add an AutogradMeta factory."


When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library. So I need another virtual factory interface
to be able to do this.

I also adjust the AutogradMeta constructor so that the TensorImpl argument is
optional. This argument is only needed if `requires_grad == True`, as we use it
to test if the variable is valid (only floating point tensors can have requires grad true).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add an AutogradMeta factory."


When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library. So I need another virtual factory interface
to be able to do this.

I also adjust the AutogradMeta constructor so that the TensorImpl argument is
optional. This argument is only needed if `requires_grad == True`, as we use it
to test if the variable is valid (only floating point tensors can have requires grad true).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171161](https://our.internmc.facebook.com/intern/diff/D18171161)

[ghstack-poisoned]

ezyang committed:
Update on "Add an AutogradMeta factory."


When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library. So I need another virtual factory interface
to be able to do this.

I also adjust the AutogradMeta constructor so that the TensorImpl argument is
optional. This argument is only needed if `requires_grad == True`, as we use it
to test if the variable is valid (only floating point tensors can have requires grad true).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171161](https://our.internmc.facebook.com/intern/diff/D18171161)

[ghstack-poisoned]

ezyang committed:
Update on "Add an AutogradMeta factory."


When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library. So I need another virtual factory interface
to be able to do this.

I also adjust the AutogradMeta constructor so that the TensorImpl argument is
optional. This argument is only needed if `requires_grad == True`, as we use it
to test if the variable is valid (only floating point tensors can have requires grad true).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171161](https://our.internmc.facebook.com/intern/diff/D18171161)

[ghstack-poisoned]

ezyang committed:
Update on "Add an AutogradMeta factory."


When I turn on Variable everywhere, I will need to be able to construct
AutogradMetas from TensorImpl.  But I cannot call the constructor directly
as it lives in another dynamic library. So I need another virtual factory interface
to be able to do this.

I also adjust the AutogradMeta constructor so that the TensorImpl argument is
optional. This argument is only needed if `requires_grad == True`, as we use it
to test if the variable is valid (only floating point tensors can have requires grad true).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171161](https://our.internmc.facebook.com/intern/diff/D18171161)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit f4104bf:
* 0/1 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=f4104bf1b7e68d9fd2b8d5a045b52a74f8dcf3d4).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).
